🐛 -e deprecated & /usr/bin -> /usr/local/bin

Rezenders opened the pull request:
I am following the [course](https://learning.edx.org/course/course-v1:DelftX+ROS1x+1T2020/block-v1:DelftX+ROS1x+1T2020+type@sequential+block@1e69cfa733964152b8d34232dedd7d2d/block-v1:DelftX+ROS1x+1T2020+type@vertical+block@a3de6a135b5843e5b0b285dff35bf8ed) and I encountered an error when running singularity in ubuntu 18.04. 

It seems that `gnome-terminal -e` is deprecated and was substituted with `gnome-terminal -- `. 

Also, when running hrwros I get this error:

```
ERROR  : Failed to mount squashfs image in (read only): Invalid argument
```
Apparently this is due to ubuntu 18.04 using the version [2.4.2](https://packages.ubuntu.com/search?suite=bionic&arch=any&searchon=names&keywords=singularity) of singulary-containter and it seems that versions <3.0.0 are not supported anymore. See: https://github.com/hpcng/singularity/issues/5482

A solution is to manually install singularity as detailed in this [link](https://github.com/hpcng/singularity/blob/master/INSTALL.md).
When building following the instructions singularity is placed in `/usr/local/bin` that is why I also proposed this change. It is also possible to manually set the place where it gets installed, so if you find better I can change the commit to keep using `/usr/bin`

Rezenders committed:
🐛 -e deprecated & /usr/bin -> /usr/local/bin

gavanderhoorn commented on start-hrwros-nvidia.desktop:
```diff
@@ -2,13 +2,18 @@
 Comment=Startup the environment for HRWROS for computers with NVIDIA graphics cards.
-Exec=gnome-terminal -e '/bin/bash -c "/bin/mkdir -p /tmp/hrwros_run/dconf && XDG_RUNTIME_DIR=/tmp/hrwros_run /usr/bin/singularity shell -p --nv $(xdg-user-dir DOWNLOAD)/hrwros-09.simg"' --class=hrwros_mooc_class_nv
 Icon=hrwros-mooc-icon-nv.png
+Exec=gnome-terminal -- bash -c "/bin/mkdir -p /tmp/hrwros_run/dconf && XDG_RUNTIME_DIR=/tmp/hrwros_run /
+usr/local/bin/singularity shell -p --nv $(xdg-user-dir DOWNLOAD)/hrwros-09.simg" --class=hrwros_mooc_cla
+ss_nv
```
I'm not sure this will work like this (ie: hard-wrapped across multiple lines).

Please revert the wrapping.

gavanderhoorn replied:
Same for the other `Exec` entries.

Rezenders replied:
ok, it was unintentional

gavanderhoorn reviewed (changes requested):
Thank you for the PR.

I've *requested changes* as I believe it would be good to keep changesets minimal, meaning: let's not fix two issues in the same commit / PR.

If you could please revert the change to the path to `singularity`, we could review the proposed change to the `gnome-terminal` args.

Rezenders committed:
🐛 fix line break & revert /usr/local/bin->/usr/bin

Rezenders reviewed (commented):
Done

gavanderhoorn commented on start-hrwros-nvidia.desktop:
```diff
@@ -2,13 +2,14 @@
 Comment=Startup the environment for HRWROS for computers with NVIDIA graphics cards.
-Exec=gnome-terminal -e '/bin/bash -c "/bin/mkdir -p /tmp/hrwros_run/dconf && XDG_RUNTIME_DIR=/tmp/hrwros_run /usr/bin/singularity shell -p --nv $(xdg-user-dir DOWNLOAD)/hrwros-09.simg"' --class=hrwros_mooc_class_nv
 Icon=hrwros-mooc-icon-nv.png
+Exec=gnome-terminal -- bash -c "/bin/mkdir -p /tmp/hrwros_run/dconf && XDG_RUNTIME_DIR=/tmp/hrwros_run /usr/bin/singularity shell -p --nv $(xdg-user-dir DOWNLOAD)/hrwros-09.simg" --class=hrwros_mooc_class_nv
```
```suggestion
Exec=gnome-terminal -- /bin/bash -c "/bin/mkdir -p /tmp/hrwros_run/dconf && XDG_RUNTIME_DIR=/tmp/hrwros_run /usr/bin/singularity shell -p --nv $(xdg-user-dir DOWNLOAD)/hrwros-09.simg" --class=hrwros_mooc_class_nv
```

gavanderhoorn commented on start-hrwros-nvidia.desktop:
```diff
@@ -2,13 +2,14 @@
 [Desktop Action new-ccs]
 Name=New CCS (Nvidia)
-Exec=gnome-terminal -e '/bin/bash -c "/bin/mkdir -p /tmp/hrwros_run/dconf && XDG_RUNTIME_DIR=/tmp/hrwros_run /usr/bin/singularity shell -p --nv $(xdg-user-dir DOWNLOAD)/hrwros-09.simg"' --class=hrwros_mooc_class_nv
+Exec=gnome-terminal -- bash -c "/bin/mkdir -p /tmp/hrwros_run/dconf && XDG_RUNTIME_DIR=/tmp/hrwros_run /usr/bin/singularity shell -p --nv $(xdg-user-dir DOWNLOAD)/hrwros-09.simg" --class=hrwros_mooc_class_nv
```
```suggestion
Exec=gnome-terminal -- /bin/bash -c "/bin/mkdir -p /tmp/hrwros_run/dconf && XDG_RUNTIME_DIR=/tmp/hrwros_run /usr/bin/singularity shell -p --nv $(xdg-user-dir DOWNLOAD)/hrwros-09.simg" --class=hrwros_mooc_class_nv
```

gavanderhoorn commented on start-hrwros.desktop:
```diff
@@ -2,14 +2,14 @@
 Comment=Startup the environment for HRWROS (non-NVIDIA computers).
-Exec=gnome-terminal -e '/bin/bash -c "/bin/mkdir -p /tmp/hrwros_run/dconf && XDG_RUNTIME_DIR=/tmp/hrwros_run /usr/bin/singularity shell -p $(xdg-user-dir DOWNLOAD)/hrwros-09.simg"' --class=hrwros_mooc_class
 Icon=hrwros-mooc-icon.png
+Exec=gnome-terminal -- bash -c "/bin/mkdir -p /tmp/hrwros_run/dconf && XDG_RUNTIME_DIR=/tmp/hrwros_run /usr/bin/singularity shell -p $(xdg-user-dir DOWNLOAD)/hrwros-09.simg" --class=hrwros_mooc_class
```
```suggestion
Exec=gnome-terminal -- /bin/bash -c "/bin/mkdir -p /tmp/hrwros_run/dconf && XDG_RUNTIME_DIR=/tmp/hrwros_run /usr/bin/singularity shell -p $(xdg-user-dir DOWNLOAD)/hrwros-09.simg" --class=hrwros_mooc_class
```

gavanderhoorn commented on start-hrwros.desktop:
```diff
@@ -2,14 +2,14 @@
 [Desktop Action new-ccs]
 Name=New CCS
-Exec=gnome-terminal -e '/bin/bash -c "/bin/mkdir -p /tmp/hrwros_run/dconf && XDG_RUNTIME_DIR=/tmp/hrwros_run /usr/bin/singularity shell -p $(xdg-user-dir DOWNLOAD)/hrwros-09.simg"' --class=hrwros_mooc_class
+Exec=gnome-terminal -- bash -c "/bin/mkdir -p /tmp/hrwros_run/dconf && XDG_RUNTIME_DIR=/tmp/hrwros_run /usr/bin/singularity shell -p $(xdg-user-dir DOWNLOAD)/hrwros-09.simg" --class=hrwros_mooc_class
```
```suggestion
Exec=gnome-terminal -- /bin/bash -c "/bin/mkdir -p /tmp/hrwros_run/dconf && XDG_RUNTIME_DIR=/tmp/hrwros_run /usr/bin/singularity shell -p $(xdg-user-dir DOWNLOAD)/hrwros-09.simg" --class=hrwros_mooc_class
```

gavanderhoorn reviewed (changes requested):
Let's keep using absolute paths.

gavanderhoorn committed:
Update start-hrwros-nvidia.desktop

gavanderhoorn committed:
Update start-hrwros-nvidia.desktop

gavanderhoorn committed:
Update start-hrwros.desktop

gavanderhoorn committed:
Update start-hrwros.desktop

gavanderhoorn commented:
@marioney: I cannot test this right now. Could you make sure the launcher still works?

Or perhaps @resmaessen?

marioney commented:
I've tested and it runs singularity but does not se the class correctly.  

If we change `-e` to `--` we need to put the `--class` command before, so it's not ignored

marioney committed:
Change location of `--class` in exec command

Updates exec command in start-hrwros-nvidia.desktop

  ` --class=hrwros_mooc_class_nv ` set before the command so it's not ignored in both main exec and new ccs action

marioney committed:
Change location of --class in exec command 

Updates exec command in start-hrwros.desktop

  ` --class=hrwros_mooc_class ` set before the command so it's not ignored in both main exec and new ccs action

marioney commented:
with the changes of https://github.com/rosin-project/mooc_hrwros_unity_launcher/pull/13/commits/7a1fde2b2defb0603a6e0ef8c1219dff6ed694e0 it looks fine for me.

@gavanderhoorn if you agree, we can merge this

gavanderhoorn commented:
thanks @Rezenders for the fix :+1: 

And thanks @marioney for testing it.

gavanderhoorn commented:
Squash-merged as this was a really trivial change and I didn't want to get all those fixup commits in the history.

gavanderhoorn commented:
As to the Singularity error: we're aware of that, but we cannot ask all students / learners to build Singularity from source.

I believe we should stick to assuming a system-installed Singularity is available and if not, the setup script should probably be changed to update the path depending on where it is found. Hard-coding a `/usr/local` location in the `.desktop` file should be avoided.

marioney commented:
I've just tested the https://github.com/rosin-project/mooc_hrwros_unity_launcher/blob/master/install-hrwros-starter.sh and it works fine :tada: 

> As to the Singularity error: we're aware of that, but we cannot ask all students / learners to build Singularity from source.

In any case, if we use the new version of singularity (3.7 ds1) the `Failed to mount squashfs image in (read only): Invalid argument` Error is no longer an issue.
